google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java

```diff
@@ -2922,6 +3014,40 @@ boolean isValid() {
     }
   }
 
+  /**
+   * Returns a multiplexed session. The method fallbacks to a regular session if {@link
+   * SessionPoolOptions#useMultiplexedSession} is not set.
+   */
+  SessionFutureWrapper getMultiplexedSessionWithFallback() throws SpannerException {
+    if (options.getUseMultiplexedSession()) {
+      ISpan span = tracer.getCurrentSpan();
+      try {
+        SessionFutureWrapper sessionFuture = getWrappedMultiplexedSessionFuture();
+        incrementNumSessionsInUse(true);
+        return sessionFuture;
+      } catch (Throwable t) {
+        span.addAnnotation("No multiplexed session available.");
+        throw SpannerExceptionFactory.asSpannerException(t.getCause());
+      }
+    } else {
+      return new PooledSessionFutureWrapper(getSession());
+    }
+  }
+
+  SessionFutureWrapper getWrappedMultiplexedSessionFuture() {
+    return new MultiplexedSessionFutureWrapper(currentMultiplexedSessionReference.get());
```

Comment:
We only need one instance of `MultiplexedSessionFutureWrapper` per instance of `SettableFuture` that we have for the multiplexed session, as the wrapper does not contain any additional state. So we should just create one instance of it per multiplexed session that we have (meaning one at startup, and one each time that we successfully refresh the multiplexed session). That will reduce the number of objects that we have to create each time a read-only operation is executed.

Reply:
Done.

Reply:
Yes, for the time-being it is stateless. Hence, stored a local variable `MultiplexedSessionFutureWrapper` and managing this in the call back methods. 

I fore-see that we may need to store the state of `span` in the Future object. I'll change this back to a stateful object based on how that refactoring goes at - https://github.com/googleapis/java-spanner/pull/3016